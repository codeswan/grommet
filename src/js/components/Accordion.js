// (C) Copyright 2014-2016 Hewlett Packard Enterprise Development LP

import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';
import List from './List';

import CSSClassnames from '../utils/CSSClassnames';

const CLASS_ROOT = CSSClassnames.ACCORDION;

export default class Accordion extends Component {
  constructor(props, context) {
    super(props, context);
    this._activatePanel = this._activatePanel.bind(this);

    this.state = {
      activeIndex: props.initialIndex
    };
  }

  _activatePanel (index) {
    this.setState({activeIndex: index});
  }

  render () {
    const {
      animate,
      className,
      children,
      openMulti,
      ...props
    } = this.props;

    const classes = classnames(
      CLASS_ROOT,
      className
    );

    const accordionChildren = React.Children
      .map(children, (child, index) => {
        return React.cloneElement(child, {
          id: 'accordion-panel-' + index,
          active: !openMulti ? (this.state.activeIndex === index)
            : child.props.active,
          onActive: () => {
            this._activatePanel(index);
          },
          animate
        });
      });

    return (
      <List role="tablist" className={classes} {...props}>
        {accordionChildren}
      </List>
    );
  }
};

Accordion.propTypes = {
  animate: PropTypes.bool,
  openMulti: PropTypes.bool,
  initialIndex: PropTypes.number
};

Accordion.defaultProps = {
  openMulti: false,
  animate: true
};
