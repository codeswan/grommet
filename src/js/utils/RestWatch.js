// (C) Copyright 2014-2016 Hewlett Packard Enterprise Development LP

import Rest from './Rest';

const RECONNECT_TIMEOUT = 5000; // 5s
const POLL_TIMEOUT = 10000; // 10s

let state = {
  ws: null,
  wsReady: false,
  timer: null,
  requests: [], // {message: , context: }
  nextRequestId: 1,
  initialized: false,
  socketUrl: null
};

export default {

  _sendMessage (op, id, url, params) {
    state.ws.send(JSON.stringify({
      op: op,
      id: id,
      url: url,
      params: params
    }));
  },

  _onOpen () {
    state.wsReady = true;
    // send any requests we have queued up
    state.requests.forEach(function(request) {
      this._sendMessage('start', request.id, request.url, request.params);
    }, this);
  },

  _onError (error) {
    console.log('!!! RestWatch _onError TODO', error);
  },

  _onMessage (event) {
    var message = JSON.parse(event.data);
    // Figure out which message this corresponds to so we
    // know which action to deliver the response with.
    state.requests.some(function(request) {
      if (request.id === message.id) {
        request.handler(message.result);
      }
    });
  },

  _onClose () {
    // lost connection, retry in a bit
    state.ws = null;
    state.wsReady = false;
    state.initialized = false;
    state.timer = setTimeout(this.initialize.bind(this), RECONNECT_TIMEOUT);
  },

  _getREST (request) {
    request.pollBusy = true;
    Rest.get(request.url, request.params)
      .end(function(err, res) {
        if (err) {
          throw err;
        }

        if (res.ok) {
          request.handler(res.body);
        }
        request.pollBusy = false;
      });
  },

  _poll () {
    state.requests.forEach(function(request) {
      if (!request.pollBusy) {
        this._getREST(request);
      }
    });
  },

  initialize (socketUrl) {
    if (!state.initialized && !state.ws && this.available() &&
      (state.socketUrl || socketUrl)) {
      state.socketUrl = state.socketUrl || socketUrl;
      state.ws = new WebSocket(state.socketUrl);
      state.ws.onopen = this._onOpen.bind(this);
      state.ws.onerror = this._onError.bind(this);
      state.ws.onmessage = this._onMessage.bind(this);
      state.ws.onclose = this._onClose.bind(this);
      state.initialized = true;
    }
  },

  available () {
    return ('WebSocket' in window || 'MozWebSocket' in window);
  },

  start (url, params, handler) {
    this.initialize();
    var request = {
      id: state.nextRequestId,
      url: url,
      params: params,
      handler: handler
    };
    state.nextRequestId += 1;
    state.requests.push(request);

    if (state.wsReady) {
      this._sendMessage('start', request.id, request.url, request.params);
    } else if (!this.available()) {
      // no web sockets, fall back to polling
      this._getREST(request);
      clearTimeout(state.timer);
      state.timer = setTimeout(this._poll.bind(this), POLL_TIMEOUT);
    }
    return request.id;
  },

  stop (requestId) {
    state.requests = state.requests.filter(function(request) {
      if (request.id === requestId) {
        if (state.wsReady) {
          this._sendMessage('stop', request.id);
        }
        return false;
      } else {
        return true;
      }
    }, this);
  }
};
