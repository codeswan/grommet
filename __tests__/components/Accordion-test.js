// (C) Copyright 2014-2016 Hewlett Packard Enterprise Development LP

import React from 'react';
import renderer from 'react/lib/ReactTestRenderer';

import Accordion from '../../src/js/components/Accordion';
import AccordionPanel from '../../src/js/components/AccordionPanel';

import { findAllByType } from '../utils/renderer-finder';

// needed because this:
// https://github.com/facebook/jest/issues/1353
jest.mock('react-dom');

describe('Accordion', () => {
  it('has correct default options', () => {
    const component = renderer.create(
      <Accordion>
        <AccordionPanel heading="First Title">
          <p>test</p>
        </AccordionPanel>
      </Accordion>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('has correct openMulti=true rendering', () => {
    const component = renderer.create(
      <Accordion openMulti>
        <AccordionPanel active heading="First Title">
          <p>test 1</p>
        </AccordionPanel>
        <AccordionPanel active heading="Second Title">
          <p>test 2</p>
        </AccordionPanel>
      </Accordion>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('has correct initialIndex=0 rendering', () => {
    const component = renderer.create(
      <Accordion initialIndex={0}>
        <AccordionPanel active heading="First Title">
          <p>test 1</p>
        </AccordionPanel>
        <AccordionPanel active heading="Second Title">
          <p>test 2</p>
        </AccordionPanel>
      </Accordion>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // activating the second panel
    const panelHeaders = findAllByType(tree, 'header');
    panelHeaders[1].props.onClick();

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
