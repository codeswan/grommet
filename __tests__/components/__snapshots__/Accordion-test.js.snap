exports[`Accordion has correct default options 1`] = `
<ul
  className="grommetux-list grommetux-accordion"
  onClick={[Function bound _onClick]}>
  <li
    className="grommetux-box grommetux-box--direction-column grommetux-box--align-center grommetux-box--responsive grommetux-box--pad-none grommetux-box--separator-bottom grommetux-list-item grommetux-accordion-panel"
    id={undefined}
    onClick={undefined}
    role={undefined}
    style={Object {}}
    tabIndex={undefined}>
    <header
      aria-label="Box"
      className="grommetux-box grommetux-box--full-horizontal grommetux-box--direction-row grommetux-box--justify-between grommetux-box--align-center grommetux-box--pad-horizontal-medium grommetux-box--pad-vertical-small grommetux-box--clickable grommetux-header grommetux-accordion-panel__header"
      id={undefined}
      onClick={[Function bound _onClickPanel]}
      role="tab"
      style={Object {}}
      tabIndex={0}>
      First Title
      <svg
        aria-labelledby="tab-next-title"
        className="grommetux-control-icon grommetux-control-icon-tab-next grommetux-accordion-panel__control"
        height="24px"
        role="img"
        version="1.1"
        viewBox="0 0 24 24"
        width="24px">
        <title
          id="tab-next-title">
          <span
            id="tab-next">
            tab-next
          </span>
        </title>
        <g>
          <rect
            fill="none"
            height="24"
            width="24"
            x="0"
            y="0" />
          <polygon
            fill="none"
            points="7.5,21.3768 7.5,3.0434 20.5,12.2101 	"
            stroke="#000000"
            strokeMiterlimit="10"
            strokeWidth="2" />
        </g>
      </svg>
    </header>
    <span
      className="grommetux-collapsible__wrapper" />
  </li>
</ul>
`;

exports[`Accordion has correct initialIndex=0 rendering 1`] = `
<ul
  className="grommetux-list grommetux-accordion"
  onClick={[Function bound _onClick]}>
  <li
    className="grommetux-box grommetux-box--direction-column grommetux-box--align-center grommetux-box--responsive grommetux-box--pad-none grommetux-box--separator-bottom grommetux-list-item grommetux-accordion-panel grommetux-accordion-panel--active"
    id={undefined}
    onClick={undefined}
    role={undefined}
    style={Object {}}
    tabIndex={undefined}>
    <header
      aria-label="Box"
      className="grommetux-box grommetux-box--full-horizontal grommetux-box--direction-row grommetux-box--justify-between grommetux-box--align-center grommetux-box--pad-horizontal-medium grommetux-box--pad-vertical-small grommetux-box--clickable grommetux-header grommetux-accordion-panel__header"
      id={undefined}
      onClick={[Function bound _onClickPanel]}
      role="tab"
      style={Object {}}
      tabIndex={0}>
      First Title
      <svg
        aria-labelledby="tab-next-title"
        className="grommetux-control-icon grommetux-control-icon-tab-next grommetux-accordion-panel__control"
        height="24px"
        role="img"
        version="1.1"
        viewBox="0 0 24 24"
        width="24px">
        <title
          id="tab-next-title">
          <span
            id="tab-next">
            tab-next
          </span>
        </title>
        <g>
          <rect
            fill="none"
            height="24"
            width="24"
            x="0"
            y="0" />
          <polygon
            fill="none"
            points="7.5,21.3768 7.5,3.0434 20.5,12.2101 	"
            stroke="#000000"
            strokeMiterlimit="10"
            strokeWidth="2" />
        </g>
      </svg>
    </header>
    <span
      className="grommetux-collapsible__wrapper">
      <div
        className="grommetux-collapsible"
        role="tabpanel">
        <p>
          test 1
        </p>
      </div>
    </span>
  </li>
  <li
    className="grommetux-box grommetux-box--direction-column grommetux-box--align-center grommetux-box--responsive grommetux-box--pad-none grommetux-box--separator-bottom grommetux-list-item grommetux-accordion-panel"
    id={undefined}
    onClick={undefined}
    role={undefined}
    style={Object {}}
    tabIndex={undefined}>
    <header
      aria-label="Box"
      className="grommetux-box grommetux-box--full-horizontal grommetux-box--direction-row grommetux-box--justify-between grommetux-box--align-center grommetux-box--pad-horizontal-medium grommetux-box--pad-vertical-small grommetux-box--clickable grommetux-header grommetux-accordion-panel__header"
      id={undefined}
      onClick={[Function bound _onClickPanel]}
      role="tab"
      style={Object {}}
      tabIndex={0}>
      Second Title
      <svg
        aria-labelledby="tab-next-title"
        className="grommetux-control-icon grommetux-control-icon-tab-next grommetux-accordion-panel__control"
        height="24px"
        role="img"
        version="1.1"
        viewBox="0 0 24 24"
        width="24px">
        <title
          id="tab-next-title">
          <span
            id="tab-next">
            tab-next
          </span>
        </title>
        <g>
          <rect
            fill="none"
            height="24"
            width="24"
            x="0"
            y="0" />
          <polygon
            fill="none"
            points="7.5,21.3768 7.5,3.0434 20.5,12.2101 	"
            stroke="#000000"
            strokeMiterlimit="10"
            strokeWidth="2" />
        </g>
      </svg>
    </header>
    <span
      className="grommetux-collapsible__wrapper" />
  </li>
</ul>
`;

exports[`Accordion has correct initialIndex=0 rendering 2`] = `
<ul
  className="grommetux-list grommetux-accordion"
  onClick={[Function bound _onClick]}>
  <li
    className="grommetux-box grommetux-box--direction-column grommetux-box--align-center grommetux-box--responsive grommetux-box--pad-none grommetux-box--separator-bottom grommetux-list-item grommetux-accordion-panel"
    id={undefined}
    onClick={undefined}
    role={undefined}
    style={Object {}}
    tabIndex={undefined}>
    <header
      aria-label="Box"
      className="grommetux-box grommetux-box--full-horizontal grommetux-box--direction-row grommetux-box--justify-between grommetux-box--align-center grommetux-box--pad-horizontal-medium grommetux-box--pad-vertical-small grommetux-box--clickable grommetux-header grommetux-accordion-panel__header"
      id={undefined}
      onClick={[Function bound _onClickPanel]}
      role="tab"
      style={Object {}}
      tabIndex={0}>
      First Title
      <svg
        aria-labelledby="tab-next-title"
        className="grommetux-control-icon grommetux-control-icon-tab-next grommetux-accordion-panel__control"
        height="24px"
        role="img"
        version="1.1"
        viewBox="0 0 24 24"
        width="24px">
        <title
          id="tab-next-title">
          <span
            id="tab-next">
            tab-next
          </span>
        </title>
        <g>
          <rect
            fill="none"
            height="24"
            width="24"
            x="0"
            y="0" />
          <polygon
            fill="none"
            points="7.5,21.3768 7.5,3.0434 20.5,12.2101 	"
            stroke="#000000"
            strokeMiterlimit="10"
            strokeWidth="2" />
        </g>
      </svg>
    </header>
    <span
      className="grommetux-collapsible__wrapper">
      <div
        className="grommetux-collapsible"
        role="tabpanel">
        <p>
          test 1
        </p>
      </div>
    </span>
  </li>
  <li
    className="grommetux-box grommetux-box--direction-column grommetux-box--align-center grommetux-box--responsive grommetux-box--pad-none grommetux-box--separator-bottom grommetux-list-item grommetux-accordion-panel grommetux-accordion-panel--active"
    id={undefined}
    onClick={undefined}
    role={undefined}
    style={Object {}}
    tabIndex={undefined}>
    <header
      aria-label="Box"
      className="grommetux-box grommetux-box--full-horizontal grommetux-box--direction-row grommetux-box--justify-between grommetux-box--align-center grommetux-box--pad-horizontal-medium grommetux-box--pad-vertical-small grommetux-box--clickable grommetux-header grommetux-accordion-panel__header"
      id={undefined}
      onClick={[Function bound _onClickPanel]}
      role="tab"
      style={Object {}}
      tabIndex={0}>
      Second Title
      <svg
        aria-labelledby="tab-next-title"
        className="grommetux-control-icon grommetux-control-icon-tab-next grommetux-accordion-panel__control"
        height="24px"
        role="img"
        version="1.1"
        viewBox="0 0 24 24"
        width="24px">
        <title
          id="tab-next-title">
          <span
            id="tab-next">
            tab-next
          </span>
        </title>
        <g>
          <rect
            fill="none"
            height="24"
            width="24"
            x="0"
            y="0" />
          <polygon
            fill="none"
            points="7.5,21.3768 7.5,3.0434 20.5,12.2101 	"
            stroke="#000000"
            strokeMiterlimit="10"
            strokeWidth="2" />
        </g>
      </svg>
    </header>
    <span
      className="grommetux-collapsible__wrapper">
      <div
        className="grommetux-collapsible"
        role="tabpanel">
        <p>
          test 2
        </p>
      </div>
    </span>
  </li>
</ul>
`;

exports[`Accordion has correct openMulti=true rendering 1`] = `
<ul
  className="grommetux-list grommetux-accordion"
  onClick={[Function bound _onClick]}>
  <li
    className="grommetux-box grommetux-box--direction-column grommetux-box--align-center grommetux-box--responsive grommetux-box--pad-none grommetux-box--separator-bottom grommetux-list-item grommetux-accordion-panel grommetux-accordion-panel--active"
    id={undefined}
    onClick={undefined}
    role={undefined}
    style={Object {}}
    tabIndex={undefined}>
    <header
      aria-label="Box"
      className="grommetux-box grommetux-box--full-horizontal grommetux-box--direction-row grommetux-box--justify-between grommetux-box--align-center grommetux-box--pad-horizontal-medium grommetux-box--pad-vertical-small grommetux-box--clickable grommetux-header grommetux-accordion-panel__header"
      id={undefined}
      onClick={[Function bound _onClickPanel]}
      role="tab"
      style={Object {}}
      tabIndex={0}>
      First Title
      <svg
        aria-labelledby="tab-next-title"
        className="grommetux-control-icon grommetux-control-icon-tab-next grommetux-accordion-panel__control"
        height="24px"
        role="img"
        version="1.1"
        viewBox="0 0 24 24"
        width="24px">
        <title
          id="tab-next-title">
          <span
            id="tab-next">
            tab-next
          </span>
        </title>
        <g>
          <rect
            fill="none"
            height="24"
            width="24"
            x="0"
            y="0" />
          <polygon
            fill="none"
            points="7.5,21.3768 7.5,3.0434 20.5,12.2101 	"
            stroke="#000000"
            strokeMiterlimit="10"
            strokeWidth="2" />
        </g>
      </svg>
    </header>
    <span
      className="grommetux-collapsible__wrapper">
      <div
        className="grommetux-collapsible"
        role="tabpanel">
        <p>
          test 1
        </p>
      </div>
    </span>
  </li>
  <li
    className="grommetux-box grommetux-box--direction-column grommetux-box--align-center grommetux-box--responsive grommetux-box--pad-none grommetux-box--separator-bottom grommetux-list-item grommetux-accordion-panel grommetux-accordion-panel--active"
    id={undefined}
    onClick={undefined}
    role={undefined}
    style={Object {}}
    tabIndex={undefined}>
    <header
      aria-label="Box"
      className="grommetux-box grommetux-box--full-horizontal grommetux-box--direction-row grommetux-box--justify-between grommetux-box--align-center grommetux-box--pad-horizontal-medium grommetux-box--pad-vertical-small grommetux-box--clickable grommetux-header grommetux-accordion-panel__header"
      id={undefined}
      onClick={[Function bound _onClickPanel]}
      role="tab"
      style={Object {}}
      tabIndex={0}>
      Second Title
      <svg
        aria-labelledby="tab-next-title"
        className="grommetux-control-icon grommetux-control-icon-tab-next grommetux-accordion-panel__control"
        height="24px"
        role="img"
        version="1.1"
        viewBox="0 0 24 24"
        width="24px">
        <title
          id="tab-next-title">
          <span
            id="tab-next">
            tab-next
          </span>
        </title>
        <g>
          <rect
            fill="none"
            height="24"
            width="24"
            x="0"
            y="0" />
          <polygon
            fill="none"
            points="7.5,21.3768 7.5,3.0434 20.5,12.2101 	"
            stroke="#000000"
            strokeMiterlimit="10"
            strokeWidth="2" />
        </g>
      </svg>
    </header>
    <span
      className="grommetux-collapsible__wrapper">
      <div
        className="grommetux-collapsible"
        role="tabpanel">
        <p>
          test 2
        </p>
      </div>
    </span>
  </li>
</ul>
`;
